include/envoy/factory/factory.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <string>
+
+#include "common/singleton/const_singleton.h"
+
+namespace Envoy {
+namespace Factory {
+
+/**
+ * List of category names for all registered factories.
+ */
+class CategoryNames {
+public:
+  const std::string AccessLoggers = "access_loggers";
+  const std::string AccessLoggerExtensionFilters = "access_logger.extension_filters";
+  const std::string Clusters = "clusters";
+
+  const std::string FiltersHttp = "filters.http";
+  const std::string FiltersListener = "filters.listener";
+  const std::string FiltersNetwork = "filters.network";
+  const std::string FiltersUdpListener = "filters.udp_listener";
+  const std::string FiltersUpstreamNetwork = "filters.upstream_network";
+
+  const std::string GrpcCredentials = "grpc_credentials";
+  const std::string HealthCheckers = "health_checkers";
+  const std::string QuicClientCodec = "quic_client_codec";
+  const std::string QuicServerCodec = "quic_server_codec";
+  const std::string ResourceMonitors = "resource_monitors";
+  const std::string RetryPriorities = "retry_priorities";
+  const std::string RetryHostPredicates = "retry_host_predicates";
+  const std::string Resolvers = "resolvers";
+  const std::string StatsSinks = "stats_sinks";
+  const std::string TlsKeyProviders = "tls.key_providers";
+  const std::string Tracers = "tracers";
+  const std::string TransportSocketsUpstream = "transport_sockets.upstream";
+  const std::string TransportSocketsDownstream = "transport_sockets.downstream";
+  const std::string TypedMetadata = "typed_metadata";
+  const std::string UdpListeners = "udp_listeners";
+
+  // Dubbo extension
+  const std::string DubboProxyFilters = "dubbo_proxy.filters";
+  const std::string DubboProxyProtocols = "dubbo_proxy.protocols";
+  const std::string DubboProxyRouteMatchers = "dubbo_proxy.route_matchers";
+  const std::string DubboProxySerializers = "dubbo_proxy.serializers";
+
+  // Thrift extension
+  const std::string ThriftProxyFilters = "thrift_proxy.filters";
+  const std::string ThriftProxyProtocols = "thrift_proxy.protocols";
+  const std::string ThriftProxyTransports = "thrift_proxy.transports";
+
+  // WASM
+  const std::string WasmNullVm = "wasm.null_vms";
```

Comment:
drive by: is it possible to not include these here as these are only used in extensions? Can they be defined in the extensions themselves?

Reply:
The ultimate goal was to have a single point of reference to point folks at for the `category` field in https://github.com/envoyproxy/envoy/pull/9301. That said, this could be done via generated docs or something like that.

What if we define a `WasmFactoryPrefix` here (same for Dubbo and Thrift), then define the factories inside the extensions?

Reply:
In general I think we should try to never reference extensions in the core code, otherwise we have an abstraction leak that is going to bite us eventually. Can we at least spec out the "right" solution here and TODO/issue it?

Reply:
I'm still trying to wrap my head around of various changes around this area, so bear with me, please.
Taking a step back it is unclear to me why extensions have this two tiered naming convention. I s the category name used anywhere? It seems to be rather haphazardly named to be machine consumable and configurations use filter names.
It also seems that in V3 we are moving toward using the type URL of configuration proto to uniquely identify the filter.
Would it work if we finish #9391 to have factories report their config proto type URL and add that to the extension list reported by the Node?

Reply:
Yeah, I think we should definitely report type URL in #9391 as well. This could be what we use after https://github.com/envoyproxy/envoy/issues/8933 lands.

Reply:
I think for purely documentation purposes I have a simpler solution that would need to be rolled back later. Let me try that.

Reply:
The only purpose of the category (AFAIK) is for printing the extensions on startup: 

https://github.com/envoyproxy/envoy/blob/80e649cb51d8dfb3ef11f18d16cc4ff441103a52/source/server/server.cc#L253-L256

It would be really optimal if we could hold the line here and not let extensions leak into the source code any more than they already do (I think there are a few cases which we need to kill).

I don't know the background of this change though, if you think this is required can you give me more context?

Reply:
I believe @yanavlasov is going to switch to using generated RST here, so let's close this out for now.